example/generators/pacts/generatorsConsumer-generatorsProvider.json

```diff
@@ -14,7 +14,7 @@
       "request": {
         "body": {
           "content": {
-            "id": 13
+            "id": null
```

Comment:
This one feels off. Do you have a bit more of an explanation on why the id has changed if the matcher is working?

Reply:
There are 2 methods:
- `integer(13)` used `Type` matching rule, with require an integer as example value. This matching rule doesn't support a generator.
- `integerV3(13)` used `Integer` matching rule. This one doesn't require example value. So it support generator.


In this example, I want to test `ProviderState` generator.

- Because `Type` doesn't support generator, I will got the (new) `MatcherNotSupportedException` exception.
- `Integer`, on the other hand, support generator, so we didn't get this exception.

So that's why I replace `integer(13)` by `integerV3(13)`

But there is 1 problem:

If I use `integerV3(13)`, the generator `ProviderState` is still not working (it's because of PHP, not Rust. See [GeneratorAwareMatcher](https://github.com/pact-foundation/pact-php/blob/ffi/src/PhpPact/Consumer/Matcher/Matchers/GeneratorAwareMatcher.php#L34))

That's why I replace `integerV3(13)` by just `integerV3()`.

That's why `13` is replaced by `null`

Reply:
You may ask: Why on the old code, `integer(13)` still work with the generator `ProviderState`?

There are 2 reason:
- I didn't do any checking in the code, just assign the `pact:generator:type = 'ProviderState'` into the final json.
- `ProviderState` generator is a bit different from other generators:
  - Other generators work on **consumer** side, so if there is example value (e.g. `13`), generator will not have any effects.
  - `ProviderState` generator work on **provider** side. The example value (e.g. `13`) will be replaced by the value from the provider state anyway.

Reply:
We can update `GeneratorAwareMatcher`, add additional check `if (null === $this->getValue() || $this->generator instanceof ProviderState) {`, but I think it's unnecessary condition.

Reply:
I actually think the additional if statement will make it so that regression and surprise cannot take place, and the API is easier to think about. Computational cost vs user brain cost; but I'll approve (you've been very patient).

Reply:
> I actually think the additional if statement will make it so that regression and surprise cannot take place...

It's a good point actually.

> regression

Unfortunately, adding a simple condition `if (null === $this->getValue() || $this->generator instanceof ProviderState) {` doesn't help

> surprise

I think we shouldn't allow these code at the first place:
- `$this->matcher->fromProviderState($this->matcher->integer(123), '${id}')`
- `$this->matcher->fromProviderState($this->matcher->integerV3(123), '${id}')`

Example value can't be used along with generator. Generator will be silently ignored without any notice. User will be surprised. So I created this PR to handle that https://github.com/pact-foundation/pact-php/pull/420